### ISSUE TITLE
CoapObserveRelation, how to reset the reRegistrationHandler's timeout?

### ISSUE THREAD

**Issue by T0bbe** [eclipse/californium#338]
I need to start by saying that I am sorry for such a stupid question, but I am unable to solve this problem, for which I have been at for some time now.

I am implementing the Observe's functionality for OSCOAP, but I get an issue when the class CoapObserveRelation wants to re-register. It wants to re-register even though the client's request have been properly ACK:ed. Also, the server is continuously sending new notifications to the client, which the client processes properly.

The CoapObserveRelation schedules a future task and I am unable to reset its timeout. How to you guys and gals reset that timer?

**In short:** CoapObserveRelation wants to re-register even though it shouldn't and I believe this is because I don't know how to reset the future task's timer.

Best regards,
Tobias

**Comment by boaks** [comment]
Which branch are you using?

**Comment by T0bbe** [comment]
UpdateToV02

**Comment by boaks** [comment]
branch 2.0.x:

```
	protected boolean onResponse(CoapResponse response) {
		if (null != response && orderer.isNew(response.advanced())) {
			current = response;
			prepareReregistration(response, 2000);
			registrationPending.set(false);
			return true;
		} else {
			return false;
		}
	}

	private void setReregistrationHandle(ScheduledFuture<?> reregistrationHandle) {
		ScheduledFuture<?> previousHandle = this.reregistrationHandle.getAndSet(reregistrationHandle);
		if (previousHandle != null) {
			previousHandle.cancel(false);
		}
	}

	private void prepareReregistration(CoapResponse response, long backoff) {
		if (!isCanceled()) {
			long timeout = response.getOptions().getMaxAge() * 1000 + backoff;
			ScheduledFuture<?> f = scheduler.schedule(new Runnable() {
	
				@Override
				public void run() {
					reregister();
				}
			}, timeout, TimeUnit.MILLISECONDS);
			setReregistrationHandle(f);
		}
	}

```

So, are you using the MAX_AGE option? Default would be 60s.
So currently, if no notify is received for 60 seconds, the reregister will be triggered.  
This seems to be related to RFC7641, 3.3.1

UpdateToV02
Is this an branch in github/californium ?

**Comment by boaks** [comment]
> It is a branch for bitbucket.org/lseitz/oscoap_californium which is a fork from californium.

Hm, do you know the last github/californium commit this is based on?

> Yes, I am. All responses and requests are defaulted to 60 seconds. The timer still timeouts after that time, even though notifications are snowing in.

Hm, sound strange. I can check it for the current 2.0.x branch, but I don't think, this occurs there.

**Comment by T0bbe** [comment]
It looks like 1.1.0 according to the Maven dependencies.

**Comment by boaks** [comment]
You may pick the commit ee0df62bba6b658e01db00345dd5c4a53050e024 from branch
https://github.com/eclipse/californium/tree/add_observe_reregistration_test
to your version.

The test "org.eclipse.californium.core.test.ObserveTest.testObserveClientNoReregister" should show,
that the timer for the reregistration is restarted on every notification.

> Jun 28, 2017 2:16:23 PM org.eclipse.californium.core.network.UdpMatcher receiveResponse
> FINER: received response NON-2.05   MID=44169, Token=83fca6aa12861f9c, OptionSet={"Observe":6, "Max-Age":10}, ""resX says client for th".. 33 bytes
> Jun 28, 2017 2:16:23 PM org.eclipse.californium.core.CoapObserveRelation prepareReregistration
> INFO: observe relation CON-GET    MID=  974, Token=83fca6aa12861f9c, OptionSet={"Observe":0, "Uri-Path":"resX"}, no payload reregister timeout 12.000
> Received 7. Notification: NON-2.05   MID=44169, Token=83fca6aa12861f9c, OptionSet={"Observe":6, "Max-Age":10}, ""resX says client for th".. 33 bytes
> Resource resX changed to "resX says client for the 8 time"
> Jun 28, 2017 2:16:25 PM org.eclipse.californium.core.network.UdpMatcher receiveResponse
> FINER: received response NON-2.05   MID=44170, Token=83fca6aa12861f9c, OptionSet={"Observe":7, "Max-Age":10}, ""resX says client for th".. 33 bytes
> Jun 28, 2017 2:16:25 PM org.eclipse.californium.core.CoapObserveRelation prepareReregistration
> INFO: observe relation CON-GET    MID=  974, Token=83fca6aa12861f9c, OptionSet={"Observe":0, "Uri-Path":"resX"}, no payload reregister timeout 12.000
> Received 8. Notification: NON-2.05   MID=44170, Token=83fca6aa12861f9c, OptionSet={"Observe":7, "Max-Age":10}, ""resX says client for th".. 33 bytes
> Resource resX changed to "resX says client for the 9 time"
> Jun 28, 2017 2:16:27 PM org.eclipse.californium.core.network.UdpMatcher receiveResponse
> FINER: received response NON-2.05   MID=44171, Token=83fca6aa12861f9c, OptionSet={"Observe":8, "Max-Age":10}, ""resX says client for th".. 33 bytes
> Jun 28, 2017 2:16:27 PM org.eclipse.californium.core.CoapObserveRelation prepareReregistration
> INFO: observe relation CON-GET    MID=  974, Token=83fca6aa12861f9c, OptionSet={"Observe":0, "Uri-Path":"resX"}, no payload reregister timeout 12.000
> Received 9. Notification: NON-2.05   MID=44171, Token=83fca6aa12861f9c, OptionSet={"Observe":8, "Max-Age":10}, ""resX says client for th".. 33 bytes
> Jun 28, 2017 2:16:39 PM org.eclipse.californium.core.network.UdpMatcher sendRequest
> FINER: Tracking open request [MID: 975, Token: 83fca6aa12861f9c]
> Jun 28, 2017 2:16:39 PM org.eclipse.californium.core.network.UdpMatcher$ExchangeObserverImpl completed
>   
>

**Comment by boaks** [comment]
You may add logging in your version in CoapClient.ObserveMessageObserverImpl.deliver to see what's happen in your case.

**Comment by T0bbe** [comment]
Okay, now I feel really stupid. How to do I make FINER logs appear in the console window? I am sitting on Eclipse. A lot of logging is done in the code, but only the SEVERE shows up in the console.

Sorry,
Tobias

**Comment by T0bbe** [comment]
One more question, the re-registration is only supposed to be sent if the NON:s stop to arrive?
Best regards,
Tobias

**Comment by boaks** [comment]
For "pure californium" (without any additional logging implementation nor logging bridge), 

https://docs.oracle.com/cd/E19717-01/819-7753/6n9m71435/index.html

you may edit the "general logging.properties" or use 
java -Djava.util.logging.config.file=[your config file] [your app]

If the application calls `CaliforniumLogger.initialize()`, this may also initialize the logging.

**Comment by boaks** [comment]
> One more question, the re-registration is only supposed to be sent if the NON:s stop to arrive?

I guess so. (It may also work for CON notifies.)

**Comment by T0bbe** [comment]
My version of the CoapObserveRelation lacks the onResponse method, but its actions are implemented in the CoapClient instead and those rows of code are used.

**Comment by boaks** [comment]
Yes, I saw, that in the OSCOAP repo it is still a older implementation. It uses `CoapObserveRelation.setCurrent()` and 
`CoapObserveRelation.prepareReregistration()` separately in `CoapClient.ObserveMessageObserverImpl`

```
@Override protected void deliver(CoapResponse response) {
			synchronized (relation) {
				if (relation.getOrderer().isNew(response.advanced())) {
					relation.setCurrent(response);
					relation.prepareReregistration(response, 2000);
					handler.onLoad(response);
				} else {
					LOGGER.log(Level.FINER, "Dropping old notification: {0}", response.advanced());
					return;
				}
			}
		}
```

Therefore I proposed in https://github.com/eclipse/californium/issues/338#issuecomment-311645092 to add the logging there.

**Comment by boaks** [comment]
So, I would add a log info before

`synchronized (relation) {`

and after that before 

`if (relation.getOrderer().isNew(response.advanced())) {`

and then after that. With that you will get aware, if the synchronized may be a issue, of if the ordering may cause the problem.

**Comment by boaks** [comment]
Could you solve the problem with the californium fork?
I would like to close this issue, I could not reproduce it with our variants.
But if you like to have more assistance, I will keep it open.

**Comment by boaks** [comment]
If the issue still exists, please open it again or create a new one.

**Comment by T0bbe** [comment]
The issue still exists, I am have been away on vacation thus my lack of response. Sorry.
